lib/minfraud/transaction.rb

```diff
@@ -65,13 +65,14 @@ class Transaction
     CONVERT_STRING_ATTRIBUTES = [:shop_id].freeze
 
     # Initializes the Transaction using parameters set from the block
-    # @raise [TransactionError] if parameters are set incorrectly
-    def initialize
+    # @raise [TransactionAttributeValidationError] if parameters are set incorrectly
+    # @raise [TransactionAttributesMissingError] if required parameters are missing
+    def initialize(strong_validation: true)
       yield self
       unless has_required_attributes?
-        raise TransactionError, 'You did not set all the required transaction attributes.'
+        raise TransactionAttributesMissingError, 'You did not set all the required transaction attributes.'
       end
-      validate_attributes
+      validate_attributes if strong_validation
```

Comment:
As it stands right now, we'll only perform validation if `strong_validation` is `true`, and if so, any inputs that fail validation will raise an error.

What if `strong_validation` could control the exception raising as opposed to the entire validation? And if `strong_validation` is `false` then the gem just logs warnings saying that the input did not pass validation?

Something along the lines of:

```ruby
def initialize(strong_validation: true)
  @strong_validation = strong_validation
  validate_attributes
end

def fail_validation(message)
  raise TransactionAttributeValidationError, message if @strong_validation

  Rails.logger.warning([MinfraudRuby] #{message})
end

def validate_string(attr_name)
  attribute = send(attr_name)
  if attribute && !attribute.instance_of?(String)
    fail_validation("Transaction.#{attr_name} must be a string")
  end
end
```

Thoughts?

Reply:
Actually, nevermind just realized we're already doing this in the Bladerunner PR

Reply:
Part of the reason it was implemented this way was to have control over the logging method outside the gem. Apps can rescue the validation error, take whatever logging action they'd like with it, and if desired create the non-validated request.

However, you're right that we could make the logging patterns more explicit in the gem itself without the need for rescuing exceptions. We could do what you're suggesting and also pass a logger into the gem to retain the flexibility in logging.

Reply:
<details><summary>Additional context from slack  </summary> 


<img width="1120" alt="Screen Shot 2021-01-21 at 3 43 00 PM" src="https://user-images.githubusercontent.com/11155207/105410145-7f921400-5bff-11eb-8a42-9e62c99aea82.png">
<img width="1125" alt="Screen Shot 2021-01-21 at 3 43 10 PM" src="https://user-images.githubusercontent.com/11155207/105410146-80c34100-5bff-11eb-978d-acf5f6662564.png">
 


</details>